### PR TITLE
Implement time metric, flushing queue, popping chunks and more

### PR DESCRIPTION
**Pull request by FelonEkonom**
Features implemented in this PR:
 - flushing queue
 - time metric
 - popping chunks
 - push_buffer_and_pop_* functions

Relates to https://github.com/membraneframework/membrane_core/issues/734

### CODE REVIEW

File: lib/membrane/timestamp_queue.ex [hunk 1]
```diff
@@ -113,27 +132,57 @@ defmodule Membrane.TimestampQueue do
...
+            valid_offset -> valid_offset
+          end)
+          |> Map.update!(:use_pts?, fn
+            nil -> buffer.dts == nil
+            valid_boolean -> valid_boolean
+          end)
+          |> Map.update!(:timestamps_qex, fn
+            nil when timestamp_queue.metric_unit == :time -> Qex.new()
+            other -> other
+          end)
```
Comment (mat-hek) [inline lib/membrane/timestamp_queue.ex]: this could be done in `increase_buffers_size`

File: lib/membrane/timestamp_queue.ex [hunk 2]
```diff
@@ -274,76 +363,212 @@ defmodule Membrane.TimestampQueue do
...
+        boundary = timestamp_queue.pause_demand_boundary
 
-      actions =
-        if pad_queue.buffers_size < boundary and old_buffers_size >= boundary,
-          do: [resume_auto_demand: pad_ref],
-          else: []
+        actions =
+          if pad_queue.buffers_size < boundary and old_buffers_size >= boundary,
+            do: [resume_auto_demand: pad_ref],
+            else: []
```
Comment (mat-hek) [inline lib/membrane/timestamp_queue.ex:506]: can't we move this check to `do_pop` so it's only done after popping all buffers?

File: lib/membrane/timestamp_queue.ex [hunk 3]
```diff
@@ -274,76 +363,212 @@ defmodule Membrane.TimestampQueue do
...
+  A buffer `b` from pad `p` is available, if all pads different than `p`
+    - either have a buffer in the queue, that is older than `b`
+    - or haven't ever had any buffer on the queue
+    - or have end of stream pushed on the queue.
+
+  An item other than a buffer is considered available if all newer buffers on the same pad are
+  available.
+
+  The returned value is a suggested actions list, a list of chunks of popped items and the updated
+  queue.
```
Comment (mat-hek) [inline lib/membrane/timestamp_queue.ex]:
```suggestion
  Works like `pop_available_items/1`, but the returned items are arranged in chunks of duration `chunk_duration`.
  
  `chunk_duration` must be passed as an option to `new/1`. The duration of each chunk may not be exactly the `chunk_duration`, but the average duration will converge to it. With that exception, only full chunks are returned.
  
  See `pop_available_items/1` for details.
```

File: lib/membrane/timestamp_queue.ex [hunk 4]
```diff
@@ -274,76 +363,212 @@ defmodule Membrane.TimestampQueue do
...
+  An item other than a buffer is considered available if all newer buffers on the same pad are
+  available.
+
+  The returned value is a suggested actions list, a list of popped items and the updated queue.
+
+  If the amount of buffers associated with any pad in the queue
+   - falls below the `pause_demand_boundary`, the suggested actions list contains
+     `t:Membrane.Action.resume_auto_demand()` actions
+   - rises above the `pause_demand_boundary`, the suggested actions list contains
+     `t:Membrane.Action.pause_auto_demand()` action.
```
Comment (mat-hek) [inline lib/membrane/timestamp_queue.ex]:
```suggestion
  The equivalent of calling `push_buffer/2` and then `pop_available_items/1`.
```

File: lib/membrane/timestamp_queue.ex [hunk 5]
```diff
@@ -274,76 +363,212 @@ defmodule Membrane.TimestampQueue do
...
+  available.
+
+  The returned value is a suggested actions list, a list of chunks of popped items and the updated
+  queue.
+
+  If the amount of buffers associated with any pad in the queue
+   - falls below the `pause_demand_boundary`, the suggested actions list contains
+     `t:Membrane.Action.resume_auto_demand()` actions
+   - rises above the `pause_demand_boundary`, the suggested actions list contains
+     `t:Membrane.Action.pause_auto_demand()` action.
```
Comment (mat-hek) [inline lib/membrane/timestamp_queue.ex]:
```suggestion
  The equivalent of calling `push_buffer/2` and then `pop_chunked/1`.
```

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane/timestamp_queue.ex (hunk 2 above)
Comment (FelonEkonom) [reply lib/membrane/timestamp_queue.ex:506]: We would have to 
1) store old `pad_queues`
2) iterate over all returned items to get set of pads that occur with a `:buffer` item
3) compare `pad_queue.buffers_size` inside old `pad_queues` and new `pad_queues`

I don't see any specific advantage of doing it after popping all buffers, TBH it was written this way some time ago but I have rewritten it to make it simpler

File: lib/membrane/timestamp_queue.ex [hunk 6]
```diff
@@ -373,8 +546,46 @@ defmodule Membrane.TimestampQueue do
...
+
   defp push_pad_on_heap(timestamp_queue, pad_ref, priority) do
     heap_item = {priority, pad_ref}
     Map.update!(timestamp_queue, :pads_heap, &Heap.push(&1, heap_item))
   end
+
+  @doc """
+  Pops all items in the proper order and closes the queue.
+
+  After being closed, queue is unable handle any new buffer/stream format/event/end of stream.
```
Comment (mat-hek) [inline lib/membrane/timestamp_queue.ex]:
```suggestion
  After being closed, nothing can be pushed to the queue anymore - a new queue should be created if needed.
```

**Review by mat-hek — APPROVED** [review]

**Pull request merged**